src/lib/libcrypto/man/engine.3
.\"	$OpenBSD: engine.3,v 1.9 2016/12/11 12:21:48 schwarze Exp $
.\"	OpenSSL a528d4f0 Oct 27 13:40:11 2015 -0400
.\"
.\" This file was written by Geoff Thorpe <geoff@openssl.org>.
.\" Copyright (c) 2002, 2004, 2007, 2015 The OpenSSL Project.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. All advertising materials mentioning features or use of this
.\"    software must display the following acknowledgment:
.\"    "This product includes software developed by the OpenSSL Project
.\"    for use in the OpenSSL Toolkit. (http://www.openssl.org/)"
.\"
.\" 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For written permission, please contact
.\"    openssl-core@openssl.org.
.\"
.\" 5. Products derived from this software may not be called "OpenSSL"
.\"    nor may "OpenSSL" appear in their names without prior written
.\"    permission of the OpenSSL Project.
.\"
.\" 6. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by the OpenSSL Project
.\"    for use in the OpenSSL Toolkit (http://www.openssl.org/)"
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
.\" EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
.\" ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
.\" OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: December 11 2016 $
.Dt ENGINE 3
.Os
.Sh NAME
.Nm ENGINE_add ,
.Nm ENGINE_by_id ,
.Nm ENGINE_finish ,
.Nm ENGINE_get_first ,
.Nm ENGINE_get_last ,
.Nm ENGINE_get_next ,
.Nm ENGINE_get_prev ,
.Nm ENGINE_init ,
.Nm ENGINE_load_builtin_engines ,
.Nm ENGINE_remove
.Nd ENGINE cryptographic module support
.Sh SYNOPSIS
.In openssl/engine.h
.Ft ENGINE *
.Fn ENGINE_get_first void
.Ft ENGINE *
.Fn ENGINE_get_last void
.Ft ENGINE *
.Fo ENGINE_get_next
.Fa "ENGINE *e"
.Fc
.Ft ENGINE *
.Fo ENGINE_get_prev
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_add
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_remove
.Fa "ENGINE *e"
.Fc
.Ft ENGINE *
.Fo ENGINE_by_id
.Fa "const char *id"
.Fc
.Ft int
.Fo ENGINE_init
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_finish
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_load_openssl void
.Ft void
.Fn ENGINE_load_dynamic void
.Ft void
.Fn ENGINE_load_cryptodev void
.Ft void
.Fn ENGINE_load_builtin_engines void
.Ft void
.Fn ENGINE_cleanup void
.Ft ENGINE *
.Fn ENGINE_get_default_RSA void
.Ft ENGINE *
.Fn ENGINE_get_default_DSA void
.Ft ENGINE *
.Fn ENGINE_get_default_ECDH void
.Ft ENGINE *
.Fn ENGINE_get_default_ECDSA void
.Ft ENGINE *
.Fn ENGINE_get_default_DH void
.Ft ENGINE *
.Fn ENGINE_get_default_RAND void
.Ft ENGINE *
.Fo ENGINE_get_cipher_engine
.Fa "int nid"
.Fc
.Ft ENGINE *
.Fo ENGINE_get_digest_engine
.Fa "int nid"
.Fc
.Ft int
.Fo ENGINE_set_default_RSA
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_DSA
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_ECDH
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_ECDSA
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_DH
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_RAND
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_ciphers
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_digests
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_default_string
.Fa "ENGINE *e"
.Fa "const char *list"
.Fc
.Ft int
.Fo ENGINE_set_default
.Fa "ENGINE *e"
.Fa "unsigned int flags"
.Fc
.Ft unsigned int
.Fn ENGINE_get_table_flags void
.Ft void
.Fo ENGINE_set_table_flags
.Fa "unsigned int flags"
.Fc
.Ft int
.Fo ENGINE_register_RSA
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_RSA
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_RSA void
.Ft int
.Fo ENGINE_register_DSA
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_DSA
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_DSA void
.Ft int
.Fo ENGINE_register_ECDH
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_ECDH
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_ECDH void
.Ft int
.Fo ENGINE_register_ECDSA
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_ECDSA
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_ECDSA void
.Ft int
.Fo ENGINE_register_DH
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_DH
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_DH void
.Ft int
.Fo ENGINE_register_RAND
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_RAND
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_RAND void
.Ft int
.Fo ENGINE_register_STORE
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_STORE
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_STORE void
.Ft int
.Fo ENGINE_register_ciphers
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_ciphers
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_ciphers void
.Ft int
.Fo ENGINE_register_digests
.Fa "ENGINE *e"
.Fc
.Ft void
.Fo ENGINE_unregister_digests
.Fa "ENGINE *e"
.Fc
.Ft void
.Fn ENGINE_register_all_digests void
.Ft int
.Fo ENGINE_register_complete
.Fa "ENGINE *e"
.Fc
.Ft int
.Fn ENGINE_register_all_complete void
.Ft int
.Fo ENGINE_ctrl
.Fa "ENGINE *e"
.Fa "int cmd"
.Fa "long i"
.Fa "void *p"
.Fa "void (*f)(void)"
.Fc
.Ft int
.Fo ENGINE_cmd_is_executable
.Fa "ENGINE *e"
.Fa "int cmd"
.Fc
.Ft int
.Fo ENGINE_ctrl_cmd
.Fa "ENGINE *e"
.Fa "const char *cmd_name"
.Fa "long i"
.Fa "void *p"
.Fa "void (*f)(void)"
.Fa "int cmd_optional"
.Fc
.Ft int
.Fo ENGINE_ctrl_cmd_string
.Fa "ENGINE *e"
.Fa "const char *cmd_name"
.Fa "const char *arg"
.Fa "int cmd_optional"
.Fc
.Ft ENGINE *
.Fn ENGINE_new void
.Ft int
.Fo ENGINE_free
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_up_ref
.Fa "ENGINE *e"
.Fc
.Ft int
.Fo ENGINE_set_id
.Fa "ENGINE *e"
.Fa "const char *id"
.Fc
.Ft int
.Fo ENGINE_set_name
.Fa "ENGINE *e"
.Fa "const char *name"
.Fc
.Ft int
.Fo ENGINE_set_RSA
.Fa "ENGINE *e"
.Fa "const RSA_METHOD *rsa_meth"
.Fc
.Ft int
.Fo ENGINE_set_DSA
.Fa "ENGINE *e"
.Fa "const DSA_METHOD *dsa_meth"
.Fc
.Ft int
.Fo ENGINE_set_ECDH
.Fa "ENGINE *e"
.Fa "const ECDH_METHOD *dh_meth"
.Fc
.Ft int
.Fo ENGINE_set_ECDSA
.Fa "ENGINE *e"
.Fa "const ECDSA_METHOD *dh_meth"
.Fc
.Ft int
.Fo ENGINE_set_DH
.Fa "ENGINE *e"
.Fa "const DH_METHOD *dh_meth"
.Fc
.Ft int
.Fo ENGINE_set_RAND
.Fa "ENGINE *e"
.Fa "const RAND_METHOD *rand_meth"
.Fc
.Ft int
.Fo ENGINE_set_STORE
.Fa "ENGINE *e"
.Fa "const STORE_METHOD *rand_meth"
.Fc
.Ft int
.Fo ENGINE_set_destroy_function
.Fa "ENGINE *e"
.Fa "ENGINE_GEN_INT_FUNC_PTR destroy_f"
.Fc
.Ft int
.Fo ENGINE_set_init_function
.Fa "ENGINE *e"
.Fa "ENGINE_GEN_INT_FUNC_PTR init_f"
.Fc
.Ft int
.Fo ENGINE_set_finish_function
.Fa "ENGINE *e"
.Fa "ENGINE_GEN_INT_FUNC_PTR finish_f"
.Fc
.Ft int
.Fo ENGINE_set_ctrl_function
.Fa "ENGINE *e"
.Fa "ENGINE_CTRL_FUNC_PTR ctrl_f"
.Fc
.Ft int
.Fo ENGINE_set_load_privkey_function
.Fa "ENGINE *e"
.Fa "ENGINE_LOAD_KEY_PTR loadpriv_f"
.Fc
.Ft int
.Fo ENGINE_set_load_pubkey_function
.Fa "ENGINE *e"
.Fa "ENGINE_LOAD_KEY_PTR loadpub_f"
.Fc
.Ft int
.Fo ENGINE_set_ciphers
.Fa "ENGINE *e"
.Fa "ENGINE_CIPHERS_PTR f"
.Fc
.Ft int
.Fo ENGINE_set_digests
.Fa "ENGINE *e"
.Fa "ENGINE_DIGESTS_PTR f"
.Fc
.Ft int
.Fo ENGINE_set_flags
.Fa "ENGINE *e"
.Fa "int flags"
.Fc
.Ft int
.Fo ENGINE_set_cmd_defns
.Fa "ENGINE *e"
.Fa "const ENGINE_CMD_DEFN *defns"
.Fc
.Ft const char *
.Fo ENGINE_get_id
.Fa "const ENGINE *e"
.Fc
.Ft const char *
.Fo ENGINE_get_name
.Fa "const ENGINE *e"
.Fc
.Ft const RSA_METHOD *
.Fo ENGINE_get_RSA
.Fa "const ENGINE *e"
.Fc
.Ft const DSA_METHOD *
.Fo ENGINE_get_DSA
.Fa "const ENGINE *e"
.Fc
.Ft const ECDH_METHOD *
.Fo ENGINE_get_ECDH
.Fa "const ENGINE *e"
.Fc
.Ft const ECDSA_METHOD *
.Fo ENGINE_get_ECDSA
.Fa "const ENGINE *e"
.Fc
.Ft const DH_METHOD *
.Fo ENGINE_get_DH
.Fa "const ENGINE *e"
.Fc
.Ft const RAND_METHOD *
.Fo ENGINE_get_RAND
.Fa "const ENGINE *e"
.Fc
.Ft const STORE_METHOD *
.Fo ENGINE_get_STORE
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_GEN_INT_FUNC_PTR
.Fo ENGINE_get_destroy_function
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_GEN_INT_FUNC_PTR
.Fo ENGINE_get_init_function
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_GEN_INT_FUNC_PTR
.Fo ENGINE_get_finish_function
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_CTRL_FUNC_PTR
.Fo ENGINE_get_ctrl_function
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_LOAD_KEY_PTR
.Fo ENGINE_get_load_privkey_function
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_LOAD_KEY_PTR
.Fo ENGINE_get_load_pubkey_function
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_CIPHERS_PTR
.Fo ENGINE_get_ciphers
.Fa "const ENGINE *e"
.Fc
.Ft ENGINE_DIGESTS_PTR
.Fo ENGINE_get_digests
.Fa "const ENGINE *e"
.Fc
.Ft const EVP_CIPHER *
.Fo ENGINE_get_cipher
.Fa "ENGINE *e"
.Fa "int nid"
.Fc
.Ft const EVP_MD *
.Fo ENGINE_get_digest
.Fa "ENGINE *e"
.Fa "int nid"
.Fc
.Ft int
.Fo ENGINE_get_flags
.Fa "const ENGINE *e"
.Fc
.Ft const ENGINE_CMD_DEFN *
.Fo ENGINE_get_cmd_defns
.Fa "const ENGINE *e"
.Fc
.Ft EVP_PKEY *
.Fo ENGINE_load_private_key
.Fa "ENGINE *e"
.Fa "const char *key_id"
.Fa "UI_METHOD *ui_method"
.Fa "void *callback_data"
.Fc
.Ft EVP_PKEY *
.Fo ENGINE_load_public_key
.Fa "ENGINE *e"
.Fa "const char *key_id"
.Fa "UI_METHOD *ui_method"
.Fa "void *callback_data"
.Fc
.Sh DESCRIPTION
These functions create, manipulate, and use cryptographic modules
in the form of
.Vt ENGINE
objects.
These objects act as containers for implementations of cryptographic
algorithms, and support a reference-counted mechanism to allow them to
be dynamically loaded in and out of the running application.
.Pp
The cryptographic functionality that can be provided by an
.Vt ENGINE
implementation includes the following abstractions:
.Pp
.Bl -bullet -compact
.It
.Vt RSA_METHOD :
for providing alternative RSA implementations
.It
.Vt DSA_METHOD , DH_METHOD , RAND_METHOD , ECDH_METHOD ,
.Vt ECDSA_METHOD , STORE_METHOD :
similarly for other OpenSSL APIs
.It
.Vt EVP_CIPHER :
potentially multiple cipher algorithms (indexed by 'nid')
.It
.Vt EVP_DIGEST :
potentially multiple hash algorithms (indexed by 'nid')
.It
key-loading: loading public and/or private EVP_PKEY keys
.El
.Ss Reference counting and handles
Due to the modular nature of the
.Nm engine
API, pointers to
.Vt ENGINE Ns s
need to be treated as handles - i.e. not only as pointers, but also
as references to the underlying
.Vt ENGINE
object.

src/usr.bin/openssl/engine.c
/* $OpenBSD: engine.c,v 1.4 2015/07/16 15:24:41 beck Exp $ */
/* Written by Richard Levitte <richard@levitte.org> for the OpenSSL
 * project 2000.
 */
/* ====================================================================
 * Copyright (c) 2000 The OpenSSL Project.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All advertising materials mentioning features or use of this
 *    software must display the following acknowledgment:
 *    "This product includes software developed by the OpenSSL Project
 *    for use in the OpenSSL Toolkit. (http://www.OpenSSL.org/)"
 *
 * 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For written permission, please contact
 *    licensing@OpenSSL.org.
 *
 * 5. Products derived from this software may not be called "OpenSSL"
 *    nor may "OpenSSL" appear in their names without prior written
 *    permission of the OpenSSL Project.
 *
 * 6. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by the OpenSSL Project
 *    for use in the OpenSSL Toolkit (http://www.OpenSSL.org/)"
 *
 * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * This product includes cryptographic software written by Eric Young
 * (eay@cryptsoft.com).  This product includes software written by Tim
 * Hudson (tjh@cryptsoft.com).
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "apps.h"

#ifndef OPENSSL_NO_ENGINE
#include <openssl/engine.h>
#include <openssl/err.h>
#include <openssl/ssl.h>

static const char *engine_usage[] = {
	"usage: engine opts [engine ...]\n",
	" -v[v[v[v]]] - verbose mode, for each engine, list its 'control commands'\n",
	"               -vv will additionally display each command's description\n",
	"               -vvv will also add the input flags for each command\n",
	"               -vvvv will also show internal input flags\n",
	" -c          - for each engine, also list the capabilities\n",
	" -t[t]       - for each engine, check that they are really available\n",
	"               -tt will display error trace for unavailable engines\n",
	" -pre <cmd>  - runs command 'cmd' against the ENGINE before any attempts\n",
	"               to load it (if -t is used)\n",
	" -post <cmd> - runs command 'cmd' against the ENGINE after loading it\n",
	"               (only used if -t is also provided)\n",
	" NB: -pre and -post will be applied to all ENGINEs supplied on the command\n",
	" line, or all supported ENGINEs if none are specified.\n",
	" Eg. '-pre \"SO_PATH:/lib/libdriver.so\"' calls command \"SO_PATH\" with\n",
	" argument \"/lib/libdriver.so\".\n",
	NULL
};

static void
identity(char *ptr)
{
	return;
}

static int
append_buf(char **buf, const char *s, int *size, int step)
{
	if (*buf == NULL) {
		*size = step;
		*buf = malloc(*size);
		if (*buf == NULL)
			return 0;
		**buf = '\0';
	}

	if (strlen(*buf) + strlen(s) >= (unsigned int) *size) {
		*size += step;
		*buf = realloc(*buf, *size);
	}
	if (*buf == NULL)
		return 0;

	if (**buf != '\0')
		strlcat(*buf, ", ", *size);
	strlcat(*buf, s, *size);

	return 1;
}

static int
util_flags(BIO * bio_out, unsigned int flags, const char *indent)
{
	int started = 0, err = 0;
	/* Indent before displaying input flags */
	BIO_printf(bio_out, "%s%s(input flags): ", indent, indent);
	if (flags == 0) {
		BIO_printf(bio_out, "<no flags>\n");
		return 1;
	}
	/*
	 * If the object is internal, mark it in a way that shows instead of
	 * having it part of all the other flags, even if it really is.
	 */
	if (flags & ENGINE_CMD_FLAG_INTERNAL) {
		BIO_printf(bio_out, "[Internal] ");
	}
	if (flags & ENGINE_CMD_FLAG_NUMERIC) {
		BIO_printf(bio_out, "NUMERIC");
		started = 1;
	}
	/*
	 * Now we check that no combinations of the mutually exclusive
	 * NUMERIC, STRING, and NO_INPUT flags have been used. Future flags
	 * that can be OR'd together with these would need to added after
	 * these to preserve the testing logic.
	 */
	if (flags & ENGINE_CMD_FLAG_STRING) {
		if (started) {
			BIO_printf(bio_out, "|");
			err = 1;
		}
		BIO_printf(bio_out, "STRING");
		started = 1;
	}
	if (flags & ENGINE_CMD_FLAG_NO_INPUT) {
		if (started) {
			BIO_printf(bio_out, "|");
			err = 1;
		}
		BIO_printf(bio_out, "NO_INPUT");
		started = 1;
	}
	/* Check for unknown flags */
	flags = flags & ~ENGINE_CMD_FLAG_NUMERIC &
	    ~ENGINE_CMD_FLAG_STRING &
	    ~ENGINE_CMD_FLAG_NO_INPUT &
	    ~ENGINE_CMD_FLAG_INTERNAL;
	if (flags) {
		if (started)
			BIO_printf(bio_out, "|");
		BIO_printf(bio_out, "<0x%04X>", flags);
	}
	if (err)
		BIO_printf(bio_out, "  <illegal flags!>");
	BIO_printf(bio_out, "\n");
	return 1;
}

static int
util_verbose(ENGINE * e, int verbose, BIO * bio_out, const char *indent)
{
	static const int line_wrap = 78;
	int num;
	int ret = 0;
	char *name = NULL;
	char *desc = NULL;
	int flags;
	int xpos = 0;
	STACK_OF(OPENSSL_STRING) * cmds = NULL;
	if (!ENGINE_ctrl(e, ENGINE_CTRL_HAS_CTRL_FUNCTION, 0, NULL, NULL) ||
	    ((num = ENGINE_ctrl(e, ENGINE_CTRL_GET_FIRST_CMD_TYPE,
			0, NULL, NULL)) <= 0)) {
		return 1;
	}
	cmds = sk_OPENSSL_STRING_new_null();

	if (!cmds)
		goto err;
	do {
		int len;
		/* Get the command input flags */
		if ((flags = ENGINE_ctrl(e, ENGINE_CTRL_GET_CMD_FLAGS, num,
			    NULL, NULL)) < 0)
			goto err;
		if (!(flags & ENGINE_CMD_FLAG_INTERNAL) || verbose >= 4) {
			/* Get the command name */
			if ((len = ENGINE_ctrl(e, ENGINE_CTRL_GET_NAME_LEN_FROM_CMD, num,
				    NULL, NULL)) <= 0)
				goto err;
			if ((name = malloc(len + 1)) == NULL)
				goto err;
			if (ENGINE_ctrl(e, ENGINE_CTRL_GET_NAME_FROM_CMD, num, name,
				NULL) <= 0)
				goto err;
			/* Get the command description */
			if ((len = ENGINE_ctrl(e, ENGINE_CTRL_GET_DESC_LEN_FROM_CMD, num,
				    NULL, NULL)) < 0)
				goto err;
			if (len > 0) {
				if ((desc = malloc(len + 1)) == NULL)
					goto err;
				if (ENGINE_ctrl(e, ENGINE_CTRL_GET_DESC_FROM_CMD, num, desc,
					NULL) <= 0)
					goto err;
			}
			/* Now decide on the output */
			if (xpos == 0)
				/* Do an indent */
				xpos = BIO_puts(bio_out, indent);
			else
				/* Otherwise prepend a ", " */
				xpos += BIO_printf(bio_out, ", ");
			if (verbose == 1) {
				/* We're just listing names, comma-delimited */
				if ((xpos > (int) strlen(indent)) &&
				    (xpos + (int) strlen(name) > line_wrap)) {
					BIO_printf(bio_out, "\n");
					xpos = BIO_puts(bio_out, indent);
				}
				xpos += BIO_printf(bio_out, "%s", name);
			} else {
				/* We're listing names plus descriptions */
				BIO_printf(bio_out, "%s: %s\n", name,
				    (desc == NULL) ? "<no description>" : desc);
				/* ... and sometimes input flags */
				if ((verbose >= 3) && !util_flags(bio_out, flags,
					indent))
					goto err;
				xpos = 0;
			}
		}
		free(name);
		name = NULL;
		free(desc);
		desc = NULL;

		/* Move to the next command */
		num = ENGINE_ctrl(e, ENGINE_CTRL_GET_NEXT_CMD_TYPE,
		    num, NULL, NULL);
	} while (num > 0);
	if (xpos > 0)
		BIO_printf(bio_out, "\n");
	ret = 1;
err:
	if (cmds)
		sk_OPENSSL_STRING_pop_free(cmds, identity);
	free(name);
	free(desc);
	return ret;
}

static void
util_do_cmds(ENGINE * e, STACK_OF(OPENSSL_STRING) * cmds,
    BIO * bio_out, const char *indent)
{
	int loop, res, num = sk_OPENSSL_STRING_num(cmds);

	if (num < 0) {
		BIO_printf(bio_out, "[Error]: internal stack error\n");
		return;
	}
	for (loop = 0; loop < num; loop++) {
		char buf[256];
		const char *cmd, *arg;
		cmd = sk_OPENSSL_STRING_value(cmds, loop);
		res = 1;	/* assume success */
		/* Check if this command has no ":arg" */
		if ((arg = strstr(cmd, ":")) == NULL) {
			if (!ENGINE_ctrl_cmd_string(e, cmd, NULL, 0))
				res = 0;
		} else {
			if ((int) (arg - cmd) > 254) {
				BIO_printf(bio_out, "[Error]: command name too long\n");
				return;
			}
			memcpy(buf, cmd, (int) (arg - cmd));
			buf[arg - cmd] = '\0';
			arg++;	/* Move past the ":" */
			/* Call the command with the argument */
			if (!ENGINE_ctrl_cmd_string(e, buf, arg, 0))
				res = 0;
		}
		if (res)
			BIO_printf(bio_out, "[Success]: %s\n", cmd);
		else {
			BIO_printf(bio_out, "[Failure]: %s\n", cmd);
			ERR_print_errors(bio_out);
		}
	}
}

int
engine_main(int argc, char **argv)
{
	int ret = 1, i;
	const char **pp;
	int verbose = 0, list_cap = 0, test_avail = 0, test_avail_noise = 0;
	ENGINE *e;
	STACK_OF(OPENSSL_STRING) * engines = sk_OPENSSL_STRING_new_null();
	STACK_OF(OPENSSL_STRING) * pre_cmds = sk_OPENSSL_STRING_new_null();
	STACK_OF(OPENSSL_STRING) * post_cmds = sk_OPENSSL_STRING_new_null();
	int badops = 1;
	BIO *bio_out = NULL;
	const char *indent = "     ";

	bio_out = BIO_new_fp(stdout, BIO_NOCLOSE);

	argc--;
	argv++;
	while (argc >= 1) {
		if (strncmp(*argv, "-v", 2) == 0) {
			if (strspn(*argv + 1, "v") < strlen(*argv + 1))
				goto skip_arg_loop;
			if ((verbose = strlen(*argv + 1)) > 4)
				goto skip_arg_loop;
		} else if (strcmp(*argv, "-c") == 0)
			list_cap = 1;
		else if (strncmp(*argv, "-t", 2) == 0) {
			test_avail = 1;
			if (strspn(*argv + 1, "t") < strlen(*argv + 1))
				goto skip_arg_loop;
			if ((test_avail_noise = strlen(*argv + 1) - 1) > 1)
				goto skip_arg_loop;
		} else if (strcmp(*argv, "-pre") == 0) {
			argc--;
			argv++;
			if (argc == 0)
				goto skip_arg_loop;
			sk_OPENSSL_STRING_push(pre_cmds, *argv);
		} else if (strcmp(*argv, "-post") == 0) {
			argc--;
			argv++;
			if (argc == 0)
				goto skip_arg_loop;
			sk_OPENSSL_STRING_push(post_cmds, *argv);
		} else if ((strncmp(*argv, "-h", 2) == 0) ||
		    (strcmp(*argv, "-?") == 0))
			goto skip_arg_loop;
		else
			sk_OPENSSL_STRING_push(engines, *argv);
		argc--;
		argv++;
	}
	/* Looks like everything went OK */
	badops = 0;
skip_arg_loop:

	if (badops) {
		for (pp = engine_usage; (*pp != NULL); pp++)
			BIO_printf(bio_err, "%s", *pp);
		goto end;
	}
	if (sk_OPENSSL_STRING_num(engines) == 0) {
		for (e = ENGINE_get_first(); e != NULL; e = ENGINE_get_next(e)) {
			sk_OPENSSL_STRING_push(engines, (char *) ENGINE_get_id(e));
		}
	}
	for (i = 0; i < sk_OPENSSL_STRING_num(engines); i++) {
		const char *id = sk_OPENSSL_STRING_value(engines, i);
		if ((e = ENGINE_by_id(id)) != NULL) {
			const char *name = ENGINE_get_name(e);
			/* Do "id" first, then "name". Easier to auto-parse. */
			BIO_printf(bio_out, "(%s) %s\n", id, name);
			util_do_cmds(e, pre_cmds, bio_out, indent);
			if (strcmp(ENGINE_get_id(e), id) != 0) {
				BIO_printf(bio_out, "Loaded: (%s) %s\n",
				    ENGINE_get_id(e), ENGINE_get_name(e));
			}
			if (list_cap) {
				int cap_size = 256;
				char *cap_buf = NULL;
				int k, n;
				const int *nids;
				ENGINE_CIPHERS_PTR fn_c;
				ENGINE_DIGESTS_PTR fn_d;
				ENGINE_PKEY_METHS_PTR fn_pk;

				if (ENGINE_get_RSA(e) != NULL
				    && !append_buf(&cap_buf, "RSA",
					&cap_size, 256))
					goto end;
				if (ENGINE_get_DSA(e) != NULL
				    && !append_buf(&cap_buf, "DSA",
					&cap_size, 256))
					goto end;
				if (ENGINE_get_DH(e) != NULL
				    && !append_buf(&cap_buf, "DH",
					&cap_size, 256))
					goto end;
				if (ENGINE_get_RAND(e) != NULL
				    && !append_buf(&cap_buf, "RAND",
					&cap_size, 256))
					goto end;

				fn_c = ENGINE_get_ciphers(e);
				if (!fn_c)
					goto skip_ciphers;
				n = fn_c(e, NULL, &nids, 0);
				for (k = 0; k < n; ++k)
					if (!append_buf(&cap_buf,
						OBJ_nid2sn(nids[k]),
						&cap_size, 256))
						goto end;

		skip_ciphers:
				fn_d = ENGINE_get_digests(e);
				if (!fn_d)
					goto skip_digests;
				n = fn_d(e, NULL, &nids, 0);
				for (k = 0; k < n; ++k)
					if (!append_buf(&cap_buf,
						OBJ_nid2sn(nids[k]),
						&cap_size, 256))
						goto end;

		skip_digests:
				fn_pk = ENGINE_get_pkey_meths(e);
				if (!fn_pk)
					goto skip_pmeths;
				n = fn_pk(e, NULL, &nids, 0);
				for (k = 0; k < n; ++k)
					if (!append_buf(&cap_buf,
						OBJ_nid2sn(nids[k]),
						&cap_size, 256))
						goto end;
		skip_pmeths:
				if (cap_buf && (*cap_buf != '\0'))
					BIO_printf(bio_out, " [%s]\n", cap_buf);

				free(cap_buf);
			}
			if (test_avail) {
				BIO_printf(bio_out, "%s", indent);
				if (ENGINE_init(e)) {
					BIO_printf(bio_out, "[ available ]\n");
					util_do_cmds(e, post_cmds, bio_out, indent);
					/*
					 * XXX hell lacks a place for people who write functions with
					 * XXX unusable return semantics.
					 */
					if (ENGINE_finish(e) != 0 ||
					    ERR_GET_REASON(ERR_peek_last_error()) ==
					    ENGINE_R_FINISH_FAILED)
						e = NULL;
				} else {
					BIO_printf(bio_out, "[ unavailable ]\n");
					if (test_avail_noise)
						ERR_print_errors_fp(stdout);
					ERR_clear_error();
				}
			}
			if ((verbose > 0) && e != NULL &&
			    !util_verbose(e, verbose, bio_out, indent))
				goto end;
			ENGINE_free(e);
		} else
			ERR_print_errors(bio_err);
	}

	ret = 0;
end:

	ERR_print_errors(bio_err);
	sk_OPENSSL_STRING_pop_free(engines, identity);
	sk_OPENSSL_STRING_pop_free(pre_cmds, identity);
	sk_OPENSSL_STRING_pop_free(post_cmds, identity);
	if (bio_out != NULL)
		BIO_free_all(bio_out);

	return (ret);
}
#endif
